include/crocoddyl/core/integrator/rk4.hpp

```diff
@@ -77,6 +77,17 @@ class IntegratedActionModelRK4Tpl : public IntegratedActionModelAbstractTpl<_Sca
   virtual void calc(const boost::shared_ptr<ActionDataAbstract>& data, const Eigen::Ref<const VectorXs>& x,
                     const Eigen::Ref<const VectorXs>& u);
 
+  /**
+   * @brief Integrate the total cost value for nodes that depends only on the state using RK4 scheme
+   *
+   * It integrates the total cost and defines the next state as the current one. This function is commonly used in the
```

Comment:
The cost is not integrated actually because it's a terminal cost. Indeed the implementation of this method should be the same for all integration schemes and could be implemented just once in the base class, couldn't it?

Reply:
At the moment it is not possible as, for instance, RK4 uses a std vector of differential data objects.
We could restructure the RK4 as:
  - data.differential (differential for 0 node)
  - data.differentials

But, this API was not convinced you last time.

I could handle this upon agreement (but in another PR). Would you change the RK4 data API for this reason?

Reply:
Done the modification of the documentation. Thanks for catching this inaccurate documentation.

I will mark this conversation as resolved. Note that the RK4 API won't be changed in this PR, but I am still looking forward for your feedback

Reply:
Yes, I agree this would mess up the RK4 data API. At the same time, not doing it could lead to a replication of code every time we add a new integration scheme. I'm not satisfied by any of these options. The clean solution (to address in another PR) would be to split the ActionModel class into two classes: RunningActionModel and TerminalActionModel. The integration schemes would then be children of RunningActionModel and would only implement `calc(data, x, u)`. The classes children of TerminalActionModel instead would only implement `calc(data, x)`, and would not perform any integration. This would also allow us to modify the OCP class API so that the last action model specified by the user be forced to be a TerminalActionModel.

Reply:
I don't think this is a good idea. We exchanged terminal nodes (both model and data) with running nodes in MPC applications.

Furthermore, this will replicate more code than in the integrators. For instance, we would need to create for each action class two action classes, i.e., one for running, another one for terminal.

Finally, this will completely break the API.

Reply:
I knew this suggestion would be widely unpopular! :)
I agree it would break the API, but that doesn't mean it's not a good design choice.

I think the exchange of terminal nodes with running nodes in MPC applications could be addressed somehow. Not sure how, but there could be a way.

I'm not sure this would replicate much code. Probably we would just need a single TerminalActionModel class implementation because the TerminalActionModel is extremely lightweight, it's basically just a cost function that depends on state only. It has no dynamics and no integration scheme. Or am I missing something?